Reimplement migrate-if-needed.sh in go

jpbetz opened the pull request:
The `migrate-if-needed.sh` script was already partially implemented in go (see the attachlease and rollback sub-dirs), but was still unnecessarily difficult to understand and test.  This closely reimplements the original logic but with improved code structure, error handling and testing.

Where possible, go code that was previously executed as separate binaries is now statically linked into a single 'migrate' go cobra CLI app, which is then thinly wrapped by`migrate-if-needed.sh`.

There are numerous additional improvements that need to be made, but will be submitted in future PRs.  This PR is focused on achieving parity with the pre-existing functionality and introducing some much needed test coverage, in particular HA cluster upgrade test coverage.

It appears that the `attachlease` and `rollback` go binaries are no longer needed as standalones and so I have consolidated them into the new `migrate` go binary. Other than that, this change aims to be 100% backward compatible.

```release-note
NONE
```

jpbetz committed:
Reimplement migrate-if-needed.sh in go

jpbetz commented:
cc @wenjiaswe

wenjiaswe commented on cluster/images/etcd/migrate-if-needed.sh:
```diff
@@ -39,240 +39,43 @@
-done
-
-echo "$(date +'%Y-%m-%d %H:%M:%S') Migration finished"
+/usr/local/bin/migrate \
```
1. Did you miss binDir in the list? (line 49 https://github.com/kubernetes/kubernetes/pull/61241/files#diff-755f39aaeeec98fb826929445011dde0)
2. Would it be better if they are in the same order as in migrateOpts struct in migrate.go? (minor)

jpbetz replied:
#1 - The default value is correct, we don't need to explicitly set it.
#2 - I don't feel strongly about it.  The flags are also in a different order in the `migrate -h` output.  We did keep the relative order of all flags and env vars throughout `migrate-if-needed.sh`, which seems like the most important place to be consistent.

wenjiaswe replied:
Sounds good.

wenjiaswe commented on cluster/images/etcd/migrate/migrate_client.go:
```diff
@@ -0,0 +1,162 @@
+// and etcdctl commands called via the shell.
+type CombinedEtcdClient struct {
+	cfg      *EtcdMigrateCfg
+	v2client clientv2.Client
```
Is v2client supposed to be a pointer as well?

jpbetz replied:
Nope, it's an interface.

cheftako commented on cluster/images/etcd/migrate/data_dir.go:
```diff
@@ -0,0 +1,165 @@
+	}
+	if !exists {
+		glog.Infof("data directory '%s' does not exist, creating it", path)
+		err := os.MkdirAll(path, 0755)
```
I realize that the script your replacing was unspecific on the permissions for the directory being created. I believe we were probably creating it with 755 but it seems that the data directory containing things like secrets should not be world readable.

jpbetz replied:
Thanks for the nudge. I'll review what exactly happens in practice (in GCE/GKE) and see if we can restrict this to 0750

mml replied:
Related question: does our umask impact this?  (I think yes.)  By default, GNU `mkdir` with no arguments uses `0777` (modified by umask).  So the "no change" change would be to use `0777` and probably add a comment asking if we can/should tighten that up.

Unexpectedly *restricting* the permissions on update could cause an outage.

cheftako commented on cluster/images/etcd/migrate/data_dir.go:
```diff
@@ -0,0 +1,165 @@
+	if err != nil {
+		return err
+	}
+	err = os.MkdirAll(backupDir, 0755)
```
Again

cheftako commented on cluster/images/etcd/migrate/rollback_v2.go:
```diff
@@ -97,43 +88,43 @@ func main() {
-		glog.Fatal(err)
+		return err
 	}
 	w.Close()
```
Are we deliberately trying to skip close() in the event of an error saving a snapshot? That seems wrong...

jpbetz replied:
Good catch, this is what we run in production today :P  Fixing.

wenjiaswe commented on cluster/images/etcd/rollback/README.md:
```diff
@@ -1,45 +0,0 @@
-# Rollback workflow
-
-Build it in this directory.
```
Are you planning to add another README file as documentation?

jpbetz replied:
We'll consolidate docs to the `cluster/images/etcd/README.md` file.  I'll do another pass on it shortly to make sure it's fleshed out.

wenjiaswe commented on cluster/images/etcd/migrate/data_dir.go:
```diff
@@ -0,0 +1,165 @@
+	return d.dir.Close()
+}
+
+// Initialize set the version.txt to the target version if the data
```
I think this part correspond to the logic in # If there is no data in DATA_DIRECTORY, this means that we are, I like the smart new way, I am just not sure how obvious the new comment is for users, would everyone know that 1. if the data directory is empty then we don't need to do migration? 2. if version.txt in etcd data folder is set to target version then there is no migration? Maybe it's actually obvious to other people, ignore me then.

wenjiaswe commented on cluster/images/etcd/migrate/data_dir_test.go:
```diff
@@ -0,0 +1,130 @@
+limitations under the License.
+*/
+
+package main
```
Do we need a test for empty data directory?

wenjiaswe commented on cluster/images/etcd/migrate/versions_test.go:
```diff
@@ -0,0 +1,68 @@
+		version    *EtcdVersionPair
+		match      bool
+	}{
+		{"3.1.2/etcd3", &EtcdVersionPair{&EtcdVersion{3, 1, 2}, storageEtcd3}, true},
```
Shall we add some more random cases, like completely out of nowhere number of etcd version and etcdstorage version, like 23.1.6 and etcd6? As well as nonserialized etcd version pair? I know the code handles these scenario, but just for completeness of the test.

jpbetz replied:
Sure, I've expanded the test coverage.

wenjiaswe reviewed (commented):
The golang implementation is definitely a lot more easy to understand. Thanks! I just have a few minor comments.

jpbetz commented:
I've added two initial go integration tests and set up make to run them inside a docker container with all the versioned etcd binaries (/usr/local/bin/etcd-3.0.17 and such). Also added tls cert generation to those tests.  This sets everything up to do extensive HA cluster tests, which I'll add shortly.  We can also remove the scripted integration tests currently hacked in to the Makefile once this go/docker integration test suite is fully fleshed out.

jpbetz commented on cluster/images/etcd/migrate/integration_test.go:
```diff
@@ -0,0 +1,237 @@
+// generateSelfSignedCertKey creates a self-signed certificate and key for the given host.
+// Host may be an IP or a DNS name
+// You may also specify additional subject alt names (either ip or dns names) for the certificate
+func generateSelfSignedCertKey(host string, alternateIPs []net.IP, alternateDNS []string) ([]byte, []byte, error) {
```
I need to decide what do to here. This is a near copy of the code under `staging/src/k8s.io/apiserver/pkg/server/options/service_testing.go`. The only difference is that `x509.ExtKeyUsageClientAuth` is set.  I wanted to avoid checking in certs and keys to git (both because they can expire and then need to be regenerated, and also because their presence in source control can raise security concerns) so instead we're generating test certs and keys here.

mml commented on cluster/images/etcd/migrate/data_dir.go:
```diff
@@ -0,0 +1,163 @@
+}
+
+// IsEmpty returns true if the data directory is entirely empty.
+func (d *DataDirectory) IsEmpty() (bool, error) {
```
`IsEmpty()` only works correctly on the first try, because each time it's called it exhausts another directory entry from `d.dir`.

jpbetz replied:
Thanks for catching @mml! I'll rewrite this to open the file only within functions where I need it opened.

mml replied:
Actually, I think there's another level, too.  Pretty sure Readdirnames() will always return `.` and `..` even on an empty dir, except `/` has no `..`.

jpbetz replied:
I'll add tests. The references I was able to find to check if a directory is empty recommended this approach, but better to be safe and double check.

jpbetz replied:
I've added checks to data_dir_test to make sure this is correct. For posterity, here's where I found the approach https://stackoverflow.com/questions/30697324/how-to-check-if-directory-on-path-is-empty.  This is one cases where I'm a annoyed golang does not offer a convenience method.

mml commented on cluster/images/etcd/migrate/data_dir.go:
```diff
@@ -0,0 +1,163 @@
+// Write creates or overwrites the contents of the version.txt file with the given EtcdVersionPair.
+func (v *VersionFile) Write(vp *EtcdVersionPair) error {
+	data := []byte(fmt.Sprintf("%s/%s", vp.version, vp.storageVersion))
+	return ioutil.WriteFile(v.path, data, 0644)
```
Similar to above, bash default behavior is `0666` minus umask.  Not sure (a) if `WriteFile` respects umask nor (b) what our default umask has been.

jpbetz replied:
@cheftako @mml, Thanks for bringing this up. I learned something new today! 
I did some reading and according to the docs golang applies the umask to whatever permissions are provided here, so the correct thing for me to do is pass in `0666` here and `0777` for the `Mkdir*` calls and let umask be applied as appropriate.

mml commented:
It may be worthwhile to run a couple integration test that verify that with umask set to `X`, the directories and files get created with permissions `Y` and `Z`.

jpbetz commented:
@mml I've added some permission checks to the integration tests. They run in a docker image with umask 0022, so with a mkdir of 0777, we check that the resulting permissions of the directory are 0755 and so forth.

jpbetz commented:
Pushed `k8s.gcr.io/etcd:3.1.12-0` with these changes applied and bumped GCE to use the image so we can properly test it.

jpbetz commented:
/retest

jpbetz commented:
/retest

jpbetz commented:
@mml @wenjiaswe @cheftako Integration tests are in place, review feedback applied and tests are passing against a k8s.gcr.io/etcd image built from this PR. Please have another look when you get some time. Thanks for the quality feedback!

wenjiaswe commented:
lgtm

jpbetz commented:
Rebased

mml commented:
/lgtm

cheftako commented on cluster/images/etcd/migrate/data_dir.go:
```diff
@@ -0,0 +1,157 @@
+
+// Initialize set the version.txt to the target version if the data
+// directory is empty. If the data directory is non-empty, no
+// version.txt file will be written since the actual version of etcd
```
Is there any danger that the if the version.txt file exists, its contents do not match the target and we will blithely proceed to non-deterministic/incorrect results?

jpbetz replied:
I'm mildly concerned that the file might get out of sync at some point. The ideal situation would be to eliminate this file altogether and instead check the data files directly to determine version. We'll likely need to make some changes to etcd to provide a simple and safe way to perform this check. I'll make note of this as something we can improve in the future.

cheftako commented on cluster/images/etcd/migrate/integration_test.go:
```diff
@@ -0,0 +1,348 @@
+
+			// Write a value to each server, read it back.
+			parallel(servers, func(server *EtcdMigrateServer) {
+				err := server.client.Put(start.version, "/registry/k1", "value1")
```
Maybe we should vary key per server. Otherwise we might hide a bug because we are writing the same error everywhere.

jpbetz replied:
Good catch. Fixed.

cheftako commented on cluster/images/etcd/migrate/integration_test.go:
```diff
@@ -0,0 +1,348 @@
+					t.Fatalf("failed to write text value: %s", err)
+				}
+
+				value, err := server.client.Get(start.version, "/registry/k1")
```
If we do write different data on each server, would it make sense to break out the read check so we can check readability of all keys on each server?

jpbetz replied:
Yes, we'll check that all test values can be read back from all the servers. Thanks!

cheftako commented on cluster/images/etcd/migrate/integration_test.go:
```diff
@@ -0,0 +1,348 @@
+
+			// Check that test value can be read back from all the servers.
+			parallel(servers, func(server *EtcdMigrateServer) {
+				value, err := server.client.Get(end.version, "/registry/k1")
```
Ditto reading all keys

cheftako commented on cluster/images/etcd/migrate/migrate_client.go:
```diff
@@ -0,0 +1,217 @@
+	if len(kvs) == 0 {
+		return "", fmt.Errorf("key not found: %s", key)
+	}
+	return string(kvs[0].Value), nil
```
if length > 1 how do you know your index 0 is the right key/value pair? Wouldn't it be better to require an exact match until we understand how to grab the correct entry in the list?

jpbetz replied:
I'll replace this check with:
```
if len(kvs) != 1 {                                                                                                                                                                                                                                        
        return "", fmt.Errorf("expected exactly one value for key %s but got ", key, len(kvs))                                                                                                                                                                
     }
```

The API is only suppose to return 0 or 1 key for a Get request like this (since it's not a range or prefix get), but you're right, we should still check for it and document what we expect.

cheftako commented on cluster/images/etcd/migrate/migrate_client.go:
```diff
@@ -0,0 +1,217 @@
+
+	glog.Infof("Attaching lease to %d entries", len(objectsResp.Kvs))
+	for _, kv := range objectsResp.Kvs {
+		_, err := v3client.KV.Put(ctx, string(kv.Key), string(kv.Value), clientv3.WithLease(lease.ID))
```
If the etcd instance if recent enough to support PrevKv, would it be worth validating that it matches the put key/value?

jpbetz replied:
Good question. This is the code we moved over from attachlease.go.  I'll go ahead and add the check-- if it fails that strongly suggests concurrent access to the etcd instance while the migration is being run, which would be very bad.

jpbetz commented on cluster/gce/upgrade-aliases.sh:
```diff
@@ -161,7 +161,7 @@ export KUBE_GCE_ENABLE_IP_ALIASES=true
 export STORAGE_BACKEND="etcd3"
 export STORAGE_MEDIA_TYPE="application/vnd.kubernetes.protobuf"
-export ETCD_IMAGE=3.1.12
+export ETCD_IMAGE=3.1.12-0
```
Good question. I've rewritten the README and added a section explaining the version numbering.

jpbetz commented:
@cheftako Feedback applied. Thanks!

jpbetz commented:
/retest

cheftako commented on cluster/images/etcd/migrate/integration_test.go:
```diff
@@ -0,0 +1,354 @@
+			// Write a value to each server, read it back.
+			parallel(servers, func(server *EtcdMigrateServer) {
+				key := fmt.Sprintf("/registry/%s", server.cfg.name)
+				value := fmt.Sprintf("value-", server.cfg.name)
```
value-%s

jpbetz replied:
Gah. Thanks for noticing this. Fixing.

cheftako commented on cluster/images/etcd/migrate/integration_test.go:
```diff
@@ -0,0 +1,354 @@
+			parallel(servers, func(server *EtcdMigrateServer) {
+				for _, s := range servers {
+					key := fmt.Sprintf("/registry/%s", s.cfg.name)
+					value := fmt.Sprintf("value-", s.cfg.name)
```
value-%s

cheftako commented on cluster/images/etcd/migrate/migrate.go:
```diff
@@ -0,0 +1,185 @@
+	if opts.name == "" {
+		hostname, err := os.Hostname()
+		if err != nil {
+			glog.Errorf("Error while getting hostname to supply default --name: %s", err)
```
%v ?

cheftako replied:
https://github.com/golang/go/wiki/CodeReviewComments#error-strings seems to suggest that %v is the standard. https://godoc.org/github.com/pkg/errors show either works but being able to extend to more easily switch to %+v seems like a very minor advantage.

jpbetz replied:
Good to know. I'll switch over to `%v`.

cheftako commented on cluster/images/etcd/migrate/migrate.go:
```diff
@@ -0,0 +1,185 @@
+
+	bundledVersions, err := ParseSupportedVersions(opts.bundledVersionString)
+	if err != nil {
+		glog.Errorf("Failed to parse --supported-versions: %s", err)
```
%v ?

cheftako commented on cluster/images/etcd/migrate/migrate_client.go:
```diff
@@ -0,0 +1,223 @@
+	}
+	return e.runEtcdctlCommand(version,
+		"snapshot", "restore", snapshotFile,
+		"--data-dir", e.cfg.dataDirectory,
```
Would be nice to be consistent about --data-dir value and --data-dir=value (In Backup() above) Ditto all other flags.

jpbetz replied:
Yes, this is an artifact from how the flags were defined inconsistently in the bash script which I didn't clean up in the translation. I'll do that now.

cheftako commented on cluster/images/etcd/migrate/versions.go:
```diff
@@ -0,0 +1,212 @@
+	patch uint64
+}
+
+// ParseEtcdVersion parses a SemVer string to an EtcdVersion.
```
This does not support SemVer pre-release versions or release versions with build metadata, which may be ok?

jpbetz replied:
I'll swap over to a SemVer library like @lavalamp suggested.

cheftako reviewed (approved):
Minor details left as comments, but may be good to fix.

lavalamp commented:
Randomly saw this in my email-- SemVer is weird and you should probably use
a semver parsing library :/

On Thu, Mar 29, 2018 at 3:58 PM Walter Fender <notifications@github.com>
wrote:

> *@cheftako* commented on this pull request.
> ------------------------------
>
> In cluster/images/etcd/migrate/versions.go
> <https://github.com/kubernetes/kubernetes/pull/61241#discussion_r178202420>
> :
>
> > +package main
> +
> +import (
> +	"fmt"
> +	"strconv"
> +	"strings"
> +)
> +
> +// EtcdVersion specifies an etcd server binaries SemVer.
> +type EtcdVersion struct {
> +	major uint64
> +	minor uint64
> +	patch uint64
> +}
> +
> +// ParseEtcdVersion parses a SemVer string to an EtcdVersion.
>
> This does not support SemVer pre-release versions or release versions with
> build metadata, which may be ok?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/61241#pullrequestreview-108229203>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglkxpgiwEItLubGu5dfkfu5TD7BFdks5tjWcngaJpZM4Ss744>
> .
>

jpbetz commented:
/retest

jpbetz commented:
@lavalamp, @cheftako SemVer library added and feedback applies. PTAL.  Thanks!

cheftako commented on cluster/images/etcd/migrate/BUILD:
```diff
@@ -4,19 +4,31 @@ load(
+    importpath = "k8s.io/kubernetes/cluster/images/etcd/migrate",
     deps = [
         "//third_party/forked/etcd221/wal:go_default_library",
+        "//vendor/github.com/blang/semver:go_default_library",
```
We now have two semver libraries and I am not sure they agree on what is a valid version number.

cheftako commented:
/lgtm

lavalamp commented:
/assign @wojtek-t

wojtek-t commented on build/build-image/cross/Dockerfile:
```diff
@@ -71,7 +71,7 @@ RUN go get golang.org/x/tools/cmd/cover \
 
 # Download and symlink etcd. We need this for our integration tests.
-RUN export ETCD_VERSION=v3.1.12; \
+RUN export ETCD_VERSION=v3.1.13; \
```
Can we please move the upgrade to 3.1.13 to a separate PR?

jpbetz replied:
Will do.

wojtek-t commented on cluster/gce/manifests/etcd.manifest:
```diff
@@ -29,14 +29,20 @@
       { "name": "INITIAL_CLUSTER",
         "value": "{{ etcd_cluster }}"
       },
+      { "name": "LISTEN_PEER_URLS",
```
Also this one - this is fixing an existing bug (and is being cherrypicked to previous branches). Can we make that a separate PR (for potential easier rollback of any of those)?

wojtek-t commented on cluster/images/etcd/migrate/migrate.go:
```diff
@@ -0,0 +1,185 @@
+	cfg := &EtcdMigrateCfg{
+		binPath:           binPath,
+		name:              name,
+		port:              18629,
```
This has to be different for event etcd and for regular etcd - currently you will get the confilct and i guess something will break.

wojtek-t reviewed (commented):
Also - please squash the commits somehow - currently they are pretty messy.

wojtek-t commented on cluster/images/etcd/migrate/migrate.go:
```diff
@@ -0,0 +1,208 @@
+		ln, err := net.Listen("tcp", fmt.Sprintf(":%d", p))
+		if err == nil {
+			glog.Infof("port %d available, using it as temporary etcd port client port for migration", p)
+			ln.Close()
```
Won't it cause problems?
Don't we need to wait a bit until port is available again after opening it?

Maybe if you close it properly not - just trying to ensure you thought about it.

jpbetz replied:
Good question. I did a bit of testing and reading and this looks safe. The `Close()` call does the right thing here and makes sure the TIME-WAIT don't prevent a subsequent listen fail.  Here's a single process example: https://play.golang.org/p/uc2LxCB6FV2. In the case here the etcd server process will listen after the port is found, so it's a slightly different situation, but my integration test are using this and all pass. https://github.com/phayes/freeport/blob/master/freeport.go also uses this approach.

lavalamp commented:
Apologies for drive by, but it's not safe to test for a port by opening and
then expect re-opening to work later. The output of a function that finds a
port should be a net.Listener--you can use that to obtain the port if
necessary (for distributing to others).

On Thu, Apr 5, 2018 at 5:17 AM Wojciech Tyczynski <notifications@github.com>
wrote:

> *@wojtek-t* commented on this pull request.
> ------------------------------
>
> In cluster/images/etcd/migrate/migrate.go
> <https://github.com/kubernetes/kubernetes/pull/61241#discussion_r179441924>
> :
>
> > +
> +	err = migrator.MigrateIfNeeded(target)
> +	if err != nil {
> +		glog.Errorf("Migration failed: %v", err)
> +		os.Exit(1)
> +	}
> +}
> +
> +// mustFindAvailablePort searches the given port range for a port that the current process can listen on.
> +func mustFindAvailablePort(rangeStart, rangeEnd uint64) uint64 {
> +	var port uint64
> +	for p := rangeStart; port == 0 && p <= rangeEnd; p++ {
> +		ln, err := net.Listen("tcp", fmt.Sprintf(":%d", p))
> +		if err == nil {
> +			glog.Infof("port %d available, using it as temporary etcd port client port for migration", p)
> +			ln.Close()
>
> Won't it cause problems?
> Don't we need to wait a bit until port is available again after opening it?
>
> Maybe if you close it properly not - just trying to ensure you thought
> about it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/61241#pullrequestreview-109684534>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglqH_1s5qlUp2vy4NSMC28OcURduKks5tlgtPgaJpZM4Ss744>
> .
>

jpbetz commented:
/retest

jpbetz commented:
@lavalamp I agree. Ideally we would just open a listener on "localhost:0" and let go find a free port.  The problem here is that we're attempting to find a free port to start an etcd server process on and we can't pass a listener.  The good news is-  1) This is always run in a docker container where the port range we're allocaing a port from should not be exposed to the host OS so we don't really need to compete with other processes for ports 2) the previous implementation uses a static port and hoped for the best, so this is strictly better than what we had before.

lavalamp commented:
I'd actually rather continue to use a static port: that makes the
resolution of the problem--should it occur--clearer (fix the docker
container).

On Thu, Apr 5, 2018 at 10:56 AM Joe Betz <notifications@github.com> wrote:

> @lavalamp <https://github.com/lavalamp> I agree. Ideally we would just
> open a listener on "localhost:0" and let go find a free port. The problem
> here is that we're attempting to find a free port to start an etcd server
> process on and we can't pass a listener. The good news is- 1) This is
> always run in a docker container where the port range we're allocaing a
> port from should not be exposed to the host OS so we don't really need to
> compete with other processes for ports 2) the previous implementation uses
> a static port and hoped for the best, so this is strictly better than what
> we had before.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/61241#issuecomment-379022759>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAngls3vyiTrUtPIcn_V2QTz4Uragx9Jks5tllrbgaJpZM4Ss744>
> .
>

jpbetz commented:
@lavalamp I've switched back to static and added a flag incase we ever want to change it from the default of 18629.  Since etcd and etcd-events run in separate docker containers, they can safely use the same port concurrently in the master kubelet.  @wojtek-t the fix to use peer port 2381 for events is in migrate-if-needed.sh.

jpbetz commented:
/retest

jpbetz commented:
@wojtek-t @cheftako All feedback applied. PTAL.

wojtek-t commented on cluster/images/etcd/migrate/migrate.go:
```diff
@@ -0,0 +1,188 @@
+
+const (
+	versionFilename        = "version.txt"
+	defaultPort     uint64 = 18629
```
Hmm - I still don't see solution for port conflicts between main etcd and event etcd.

I think you should at least set "--port" flag when you're calling "migrate binary in the script.

jpbetz replied:
No need. Since etcd and etcd events are run in separate docker containers, and neither of them expose 18629 to the host operating system, they can concurrently use this port within the docker containers.

jpbetz replied:
@wojtek-t Ohh. Thanks so much for catching this!  I've populated the `--port` flag in `migrate-if-needed.sh`.

wojtek-t commented:
/approve no-issue

I'm fine with that - not LGTM-ing for now to give others one more change.

wojtek-t commented:
/retest

wojtek-t commented:
I'm LGTM-ing it to unblock it and have more soaking time of it before release (though it probably doesn't really matter).

wojtek-t commented:
/lgtm
